src/read/macros.rs

```diff
@@ -194,10 +195,351 @@ impl<R: Reader> fallible_iterator::FallibleIterator for DebugMacInfoIterator<R>
+        let mut input = self.section.clone();
+        input.skip(offset.0)?;
+        let header = MacroUnitHeader::parse(&mut input)?;
+        Ok(DebugMacroIterator { input, header })
+    }
+}
+
+impl<T> DebugMacro<T> {
+    /// Create a `DebugMacro` section that references the data in `self`.
+    ///
+    /// This is useful when `R` implements `Reader` but `T` does not.
+    ///
+    /// Used by `DwarfSections::borrow`.
+    pub fn borrow<'a, F, R>(&'a self, mut borrow: F) -> DebugMacro<R>
+    where
+        F: FnMut(&'a T) -> R,
+    {
+        borrow(&self.section).into()
+    }
+}
+
+impl<R> Section<R> for DebugMacro<R> {
+    fn id() -> SectionId {
+        SectionId::DebugMacro
+    }
+
+    fn reader(&self) -> &R {
+        &self.section
+    }
+}
+
+impl<R> From<R> for DebugMacro<R> {
+    fn from(macro_section: R) -> Self {
+        DebugMacro {
+            section: macro_section,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct MacroUnitHeader<R: Reader> {
+    /// The version of the macro unit header. At the moment only version 5 is defined.
+    _version: u16,
+    flags: u8,
+    debug_line_offset: DebugStrOffsetsBase<R::Offset>,
+}
+
+impl<R: Reader> MacroUnitHeader<R> {
+    const OFFSET_SIZE_FLAG: u8 = 0b0000_0001;
+    const DEBUG_LINE_OFFSET_FLAG: u8 = 0b0000_0010;
+    const OPCODE_OPERANDS_TABLE_FLAG: u8 = 0b0000_0100;
+
+    fn parse(input: &mut R) -> Result<Self> {
+        let version = input.read_u16()?;
+        let flags = input.read_u8()?;
+        let format = if flags & Self::OFFSET_SIZE_FLAG == 0 {
+            Format::Dwarf32
+        } else {
+            Format::Dwarf64
+        };
+        let debug_line_offset = if flags & Self::DEBUG_LINE_OFFSET_FLAG != 0 {
+            DebugStrOffsetsBase(input.read_offset(format)?)
+        } else {
+            DebugStrOffsetsBase(R::Offset::from_u64(0)?)
+        };
+        // if the opcode operands table flag is set, there is a table in the header which currently isn't parsed
+        if flags & Self::OPCODE_OPERANDS_TABLE_FLAG != 0 {
+            return Err(Error::UnsupportedOpcodeOperandsTable);
+        }
+        Ok(MacroUnitHeader {
+            _version: version,
+            flags,
+            debug_line_offset,
+        })
+    }
+
+    fn format(&self) -> Format {
+        if self.flags & Self::OFFSET_SIZE_FLAG == 0 {
+            Format::Dwarf32
+        } else {
+            Format::Dwarf64
+        }
+    }
+}
+
+/// an Entry in the `.debug_macro` section.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum DebugMacroItem<R, Offset = <R as Reader>::Offset>
+where
+    R: Reader<Offset = Offset>,
+    Offset: ReaderOffset,
+{
+    /// A macro definition.
+    Define {
+        /// The line number where the macro is defined.
+        line: u64,
+        /// The text of the macro: The name of the macro followed immediately by any formal
+        /// parameters including the surrounding parentheses, followed by the macro definition.
+        text: R,
```

Comment:
Instead of having `Define`, `DefineStrp`, `DefineStrx`, and `DefineSup`, can we combine them into a single `Define` variant that has `text: MacroString`, where `MacroString` is an enum for the various ways that a string can be encoded. Similarly for `Undef`.